src/celeritas/em/msc/UrbanMscScatter.hh

```diff
@@ -159,36 +157,115 @@ UrbanMscScatter::UrbanMscScatter(UrbanMscRef const& shared,
 
-    skip_sampling_ = true;
-    if (true_path_ < range_ && true_path_ > params_.geom_limit)
+    skip_sampling_ = [this, &helper] {
+        if (true_path_ == range_)
+        {
+            // Range-limited step (particle stops)
+            // TODO: probably redundant with low 'end energy'
+            return true;
+        }
+        if (true_path_ < params_.geom_limit)
+        {
+            // Very small step (NOTE: with the default values in UrbanMscData,
+            // this is redundant witih the tau_small comparison below if MFP >=
+            // 0.005 cm)
+            return true;
+        }
+
+        // Lazy calculation of end energy
+        end_energy_ = value_as<Energy>(helper.calc_end_energy(true_path_));
+
+        if (Energy{end_energy_} < params_.min_sampling_energy())
+        {
+            // Ending energy is very low
+            return true;
+        }
+        if (true_path_ <= lambda_ * params_.tau_small)
+        {
+            // Very small MFP travelled
+            return true;
+        }
+        return false;
+    }();
+
+    // TODO: there are several different sampling strategies for angle change:
+    // - very small step/very low energy endpoint: no scattering
+    // - very small mfp: (probably impossible because of condition above):
+    //   forward scatter
+    // - very large mfp: exiting angle is isotropic
+    // - large energy loss: "simple_scattering"
+
+    if (!skip_sampling_)
     {
-        end_energy_ = value_as<Energy>(helper_.calc_end_energy(true_path_));
-        skip_sampling_
-            = (end_energy_ < value_as<Energy>(params_.min_sampling_energy())
-               || true_path_ <= shared.params.limit_min_fix()
-               || true_path_ < lambda_ * params_.tau_small);
+        // Calculate number of mean free paths traveled
+        tau_ = true_path_ / [this, &helper] {
+            // Calculate the average MFP assuming the cross section varies
+            // linearly over the step
+            real_type lambda_end = helper.calc_msc_mfp(Energy{end_energy_});
+            if (std::fabs(lambda_ - lambda_end) < lambda_ * real_type(0.01))
+            {
+                // Cross section is almost constant over the step: avoid
+                // numerical explosion
+                return lambda_;
+            }
+            return (lambda_ - lambda_end) / std::log(lambda_ / lambda_end);
+        }();
+
+        if (CELER_UNLIKELY(tau_ < params_.tau_small))
+        {
+            // Small MFP travelled
+            // TODO: this should be virtually impossible because of
+            // skip_sampling_ above
+            is_displaced_ = false;
+        }
+        else if (tau_ < params_.tau_big)
+        {
+            // TODO: if displacement is disabled (because of being near the
+            // boundary or having a short step) then limit_min_ will be zero,
+            // and some corrections will not be applied.
+            // CELER_ASSERT(limit_min_ > 0);
```

Comment:
@whokion I noticed this unexpected behavior: in `UrbanMscStepLimit` we skip the calculation of `limit_min` (when entering a volume) for short steps. That means the initial step through a volume has `limit_min = 0` so `compute_theta0` doesn't account for the minimum limit. Fixing this doesn't seem to change any of the unit tests, though, so it's probably not important.

Reply:
Okay, then should we initialize limit_min_ with  something like`limit_min_fix() `as [this if block](https://github.com/sethrj/celeritas/blob/msc-units/src/celeritas/em/msc/UrbanMscStepLimit.hh#L143) may not completely cover all possible cases?

Reply:
As looking at [this line](https://geant4.kek.jp/lxr/source/processes/electromagnetic/standard/src/G4UrbanMscModel.cc#L97) or [this line](https://geant4.kek.jp/lxr/source/processes/electromagnetic/standard/src/G4UrbanMscModel.cc#L434) again, I confirm that the limit_min_ should be initialized with UrbanMscParameters::limit_min() which is apparently removed from this PR, but equivalently can be set to 10*limit_min_fix().

Reply:
OK, I removed `limit_min` because it wasn't used anywhere: but I guess it should be 😄 so we now have the same behavior as Geant4.